[llvm-debuginfo-analyzer] Fix ODR violation in llvm::logicalview::LVObject

jalopezg-git opened the pull request:
Some data members are only part of a class definition in a Debug build, e.g. `LVObject::ID`.  If `debuginfologicalview` is used as a library, `NDEBUG` cannot be used for this purpose, as this PP macro may have a different definition in a downstream project, which in turn triggers an ODR violation.  Fix it by
- Making `LVObject::ID` an unconditional data member.
- Making `LVObject::dump()` non-virtual.  Rationale: `virtual` is not needed (and it calls `print()`, which is virtual anyway).

Fixes #139098.

dwblaikie commented:
I think we already have a separate macro for this - ABI_BREAKING_CHECKS, perhaps?

jalopezg-git commented:
> I think we already have a separate macro for this - ABI_BREAKING_CHECKS, perhaps?

Hmmm... unsure; according to [llvm/docs/ProgrammersManual.rst](https://github.com/llvm/llvm-project/blob/main/llvm/docs/ProgrammersManual.rst), this PP macro serves a slightly different purpose, but we could probably also use it here.  WDYT, @CarlosAlbertoEnciso?

jalopezg-git commented:
> Oh, the issue is these aren't checks? They're only for debug dumping features? That is a bit awkward. Do they have to be ABI breaking to provide debugging features/could they be implemented in some other way? (& I don't think we enable dump functions based on whether it's a debug build, right? There's some separate macro for that?)

Usually `dump()` functions are kept in most (all?) other LLVM sub-projects, e.g. in Clang and MLIR.  We could perhaps do the same without too much penalty.

However, here there is also `LVObject::ID`, which is only used for debugging, AFAIU.  There's also a size concern (that all `LVObject` should be kept to minimum size), and this data member is not needed in non-Debug builds.

Yep, we could also raise a thread on Discourse for further discussion :+1:.

compnerd commented:
> I think we already have a separate macro for this - ABI_BREAKING_CHECKS, perhaps?

The `ABI_BREAKING_CHECKS` is meant for changes to the data structure layout that changes between Asserts and NoAsserts builds IIRC. The dump methods can be enabled/disabled separately (similar to how building with different targets is not considered ABI breaking).

I don't understand the motivation of this change. I realize that `NDEBUG` is a confusing macro, but this is not about a _debug_ build. `NDEBUG` indicates whether assertions are enabled or not.

jalopezg-git commented:
> I don't understand the motivation of this change. I realize that `NDEBUG` is a confusing macro, but this is not about a _debug_ build. `NDEBUG` indicates whether assertions are enabled or not.

Basically, when using `debuginfologicalview` as a library from a downstream project, `NDEBUG` may be defined if the downstream project is built in Release mode.
This results in a different class layout (e.g. for `LVObject`), given that, by the time the `LVObject.h` is parsed in a downstream project, `NDEBUG` may be defined (or not) regardless of whether it was defined (or not) when LLVM was built.

Thus, `NDEBUG` cannot be relied for anything that affects class layout (at least on public headers).  Looking at @dwblaikie and @compnerd comments, perhaps using `LLVM_ENABLE_ABI_BREAKING_CHECKS` (instead of introducing a new `LLVM_BUILD_DEBUG` PP macro) is the way to go :thinking:  (?)

compnerd commented:
> > I don't understand the motivation of this change. I realize that `NDEBUG` is a confusing macro, but this is not about a _debug_ build. `NDEBUG` indicates whether assertions are enabled or not.
> 
> Basically, when using `debuginfologicalview` as a library from a downstream project, `NDEBUG` may be defined if the downstream project is built in Release mode. This results in a different class layout (e.g. for `LVObject`), given that, by the time the `LVObject.h` is parsed in a downstream project, `NDEBUG` may be defined (or not) regardless of whether it was defined (or not) when LLVM was built.
> 
> Thus, `NDEBUG` cannot be relied for anything that affects class layout (at least on public headers). Looking at @dwblaikie and @compnerd comments, perhaps using `LLVM_ENABLE_ABI_BREAKING_CHECKS` (instead of introducing a new `LLVM_BUILD_DEBUG` PP macro) is the way to go 🤔 (?)

Well, `dump` shouldn't be part of the public interface, so removal is a better option. Note the guard - `LLVM_ENABLE_DUMP` is an alternative spelling. Depending on `NDEBUG` is acceptable here. This doesn't change the object layout, but does change the API surface. If the downstream project is attempting to use the `dump` methods, it should be building LLVM with `LLVM_ENABLE_DUMP`.

CarlosAlbertoEnciso commented:
> > > I don't understand the motivation of this change. I realize that `NDEBUG` is a confusing macro, but this is not about a _debug_ build. `NDEBUG` indicates whether assertions are enabled or not.
> > 
> > 
> > Basically, when using `debuginfologicalview` as a library from a downstream project, `NDEBUG` may be defined if the downstream project is built in Release mode. This results in a different class layout (e.g. for `LVObject`), given that, by the time the `LVObject.h` is parsed in a downstream project, `NDEBUG` may be defined (or not) regardless of whether it was defined (or not) when LLVM was built.
> > Thus, `NDEBUG` cannot be relied for anything that affects class layout (at least on public headers). Looking at @dwblaikie and @compnerd comments, perhaps using `LLVM_ENABLE_ABI_BREAKING_CHECKS` (instead of introducing a new `LLVM_BUILD_DEBUG` PP macro) is the way to go 🤔 (?)
> 
> Well, `dump` shouldn't be part of the public interface, so removal is a better option. Note the guard - `LLVM_ENABLE_DUMP` is an alternative spelling. Depending on `NDEBUG` is acceptable here. This doesn't change the object layout, but does change the API surface. If the downstream project is attempting to use the `dump` methods, it should be building LLVM with `LLVM_ENABLE_DUMP`.

May be some clarification:
```
  virtual void print(raw_ostream &OS, bool Full = true) const;

#if !defined(NDEBUG) || defined(LLVM_ENABLE_DUMP)
  virtual void dump() const { print(dbgs()); }
#endif
```
`dump` should be called only for _debug_ builds.
`print` can be called for both builds (_debug_ and _non debug_).
```
- Debug build:      print(dbgs())
- Non Debug Build:  print(outs());
```
The downstream project should use `print`.

jalopezg-git commented:
> Well, `dump` shouldn't be part of the public interface, so removal is a better option. Note the guard - `LLVM_ENABLE_DUMP` is an alternative spelling. Depending on `NDEBUG` is acceptable here. This doesn't change the object layout, but does change the API surface. If the downstream project is attempting to use the `dump` methods, it should be building LLVM with `LLVM_ENABLE_DUMP`.

To add to @CarlosAlbertoEnciso comment, the original purpose of this PR was to make the `LVObject` class definition consistent (across both LLVM and downstream projects); specifically,
- `LVObject::ID` (see [LVObject.h:174](https://github.com/llvm/llvm-project/pull/140265/files#diff-85d40304e38f0ab619a00f93fc3b5f5b04aa1e795319fab5589d2b06dc243326R174) non-static data member is conditionally added depending on `NDEBUG`, which clearly violates the ODR if `NDEBUG` has a different definition when building a downstream project (actually, it that is fairly common).
- While doing this change, I also noticed that `LVObject::dump()` is declared `virtual`.  I believe that, having it or not, changes how the vtable looks, and thus I included it in this patchset.

compnerd commented:
> May be some clarification:
> 
> ```
>   virtual void print(raw_ostream &OS, bool Full = true) const;
> 
> #if !defined(NDEBUG) || defined(LLVM_ENABLE_DUMP)
>   virtual void dump() const { print(dbgs()); }
> #endif
> ```
> 
> `dump` should be called only for _debug_ builds. `print` can be called for both builds (_debug_ and _non debug_).

I think that you are misreading the condition:

`!defined(NDEBUG)`
You are focusing on this rather than the rest of the condition:

`|| defined(LLVM_ENABLE_DUMP)`
You can always build LLVM with `-D LLVM_ENABLE_DUMP` being passed to CMake to enable the `dump` API if you want.


> ```
> - Debug build:      print(dbgs())
> - Non Debug Build:  print(outs());
> ```
> 
> The downstream project should use `print`.

This is also an option that is available to downstream projects.

> * `LVObject::ID` (see [LVObject.h:174](https://github.com/llvm/llvm-project/pull/140265/files#diff-85d40304e38f0ab619a00f93fc3b5f5b04aa1e795319fab5589d2b06dc243326R174)) non-static data member is conditionally added depending on `NDEBUG`, which clearly violates the ODR if `NDEBUG` has a different definition when building a downstream project (actually, that is fairly common).

That is something to guard with `LLVM_ENABLE_ABI_BREAKING_CHANGES` as that is changing the object layout itself between release and debug builds. However, the returning of the value between `0` and `ID` is something that can remain as `NDEBUG` only.

jalopezg-git commented:
> That is something to guard with `LLVM_ENABLE_ABI_BREAKING_CHANGES` as that is changing the object layout itself between release and debug builds. However, the returning of the value between `0` and `ID` is something that can remain as `NDEBUG` only.

LGTM; with `LLVM_BUILD_DEBUG`, I tried to mimic the most probable original intent of `NDEBUG`, but in a way that doesn't break downstream projects.
Going for `LLVM_ENABLE_ABI_BREAKING_CHECKS` looks also good and I can change it now.  Do we have general consensus?

jalopezg-git commented:
> > That is something to guard with `LLVM_ENABLE_ABI_BREAKING_CHANGES` as that is changing the object layout itself between release and debug builds. However, the returning of the value between `0` and `ID` is something that can remain as `NDEBUG` only.
> 
> LGTM; with `LLVM_BUILD_DEBUG`, I tried to mimic the most probable original intent of `NDEBUG`, but in a way that doesn't break downstream projects. Going for `LLVM_ENABLE_ABI_BREAKING_CHECKS` looks also good and I can change it now. Do we have general consensus?

I have dropped the original patch and went for
- Regarding `LVObject::ID`, temporarily went for `#if !defined(NDEBUG) && LLVM_ENABLE_ABI_BREAKING_CHECKS` (instead of plain `#ifndef NDEBUG`).  This should fix the original issue.  
@CarlosAlbertoEnciso and I are still discussing whether `ID` should be made unconditional (and smaller; currently `uint64_t`) instead.
- Making `LVObject::dump()` non-virtual.  Rationale: `virtual` is not needed (and it calls `print()`, which is virtual anyway.

compnerd reviewed (commented):
The changes for `dump` I think make sense. I'm not fully convinced about the remainder of the change. I think that we do want the ID checking in the debug builds (it is meant to be a debugging tool). If we are building without assertions, then we can enable it with ABI breaking changes. That is, I think that `#if !defined(NDEBUG) || ENABLE_ABI_BREAKING_CHANGES` is the right approach.

jalopezg-git commented:
> The changes for `dump` I think make sense. I'm not fully convinced about the remainder of the change. I think that we do want the ID checking in the debug builds (it is meant to be a debugging tool). If we are building without assertions, then we can enable it with ABI breaking changes. That is, I think that `#if !defined(NDEBUG) || ENABLE_ABI_BREAKING_CHANGES` is the right approach.

Thanks for the review, @compnerd!

Note that, although the pull-request mentions `[llvm-debuginfo-analyzer]`, the issue is really in `DebugInfoLogicalView` (the underlying library used by `llvm-debuginfo-analyzer`).  As described in https://github.com/llvm/llvm-project/issues/139098, the `LVObject` class layout may differ if the `NDEBUG` PP macro has different definition (other than it had when LLVM was built) in a downstream project that consumes the library.
Because in a downstream project, `NDEBUG` is usually used to indicate a Release build, using `#if !defined(NDEBUG) || ENABLE_ABI_BREAKING_CHANGES`, may still cause issues from the PoV of the consumer project.

jalopezg-git committed:
[llvm-debuginfo-analyzer] Make LVObject::dump() non-virtual (NFC)

`LVObject::dump()` does not need to be virtual as it relays to
`LVObject::print()` that is.

jalopezg-git committed:
[llvm-debuginfo-analyzer][debuginfologicalview] Fix ODR violation in LVObject

Fix ODR violation in `LVObject` when `DebugInfoLogicalView` library
is used in a downstream project.

Depending solely on `NDEBUG` for exposing some data members may cause
trouble in downstream projects.
Enable such cases only if `LLVM_ENABLE_ABI_BREAKING_CHECKS`.

jalopezg-git committed:
[llvm-debuginfo-analyzer][debuginfologicalview] Make LVObject::ID unconditional

Make the `LVObject::ID` member unconditional on `NDEBUG` or
`LLVM_ENABLE_ABI_BREAKING_CHECKS`.
Changing size of `ScopeLevel` (uint16_t) and `ID` (uint32_t) and careful reorder
of data members in LVObject (avoding unneeded padding) leaves `sizeof(LVObject)`
unchanged.

compnerd reviewed (approved):
Taking advantage of the padding sounds good to me.
